torchtune/modules/attention_utils.py

```diff
@@ -0,0 +1,243 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import logging
+from typing import Callable, List, Optional, Union
+
+import torch
+
+from torch import nn
+from torchtune.utils._version import torch_version_ge
+from torchtune.utils.logging import get_logger, log_once
+
+# We can only use flex attention / BlockMask if torch version >= 2.5.0 and GPU is Turing / SM75 and above
+_SUPPORTS_FLEX_ATTENTION = torch_version_ge(
+    "2.5.0"
+) and torch.cuda.get_device_capability() >= (7, 5)
+
+if _SUPPORTS_FLEX_ATTENTION:
+    from torch.nn.attention.flex_attention import (
+        BlockMask,
+        create_block_mask as create_block_causal_mask_flex,
+        flex_attention,
+    )
+
+    flex_attention_compiled = torch.compile(flex_attention, dynamic=False)
+
+    # We cannot do nested compile, but flex attention only has perf benefits
+    # when compiled. To insulate it from the compiler, we wrap it with
+    # compiler.disable so that it can be used regardless of whether the model
+    # is compiled or not, and flex attention always remains compiled.
+    @torch.compiler.disable(recursive=False)
```

Comment:
So where did we land on this? Recursive is not necessary, right?

Reply:
it works either way. I think it makes sense to keep recursive just to be overly cautious